WalletWasabi.Fluent.Generators/AnalyzerExtensions.cs

```diff
@@ -0,0 +1,114 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace WalletWasabi.Fluent.Generators;
+
+public static class AnalyzerExtensions
+{
+	public static List<IdentifierNameSyntax> GetUiContextReferences(this SyntaxNode node, SemanticModel semanticModel)
+	{
+		return
+			node.DescendantNodes()
+			 .OfType<IdentifierNameSyntax>()
+			 .Where(x => x.Identifier.ValueText == "UiContext")                                                   // faster verification
+			 .Where(x => semanticModel.GetTypeInfo(x).Type?.ToDisplayString() == UiContextAnalyzer.UiContextType) // slower, but safer. Only runs if previous verification passed.
+			 .ToList();
+	}
+
+	public static bool IsPrivate(this ConstructorDeclarationSyntax node)
+	{
+		return node.Modifiers.Any(m => m.IsKind(SyntaxKind.PrivateKeyword));
+	}
+
+	public static bool IsPublic(this ConstructorDeclarationSyntax node)
+	{
+		return node.Modifiers.Any(m => m.IsKind(SyntaxKind.PublicKeyword));
+	}
+
+	public static bool IsSourceGenerated(this SyntaxNode node)
+	{
+		var filePath = node.SyntaxTree.FilePath;
+
+		return filePath is null ||
+			   filePath.EndsWith(UiContextAnalyzer.UiContextFileSuffix);
+	}
+
+	public static bool IsSubTypeOf(this SyntaxNode node, SemanticModel model, string baseType)
```

Comment:
Yeah this code uses `SyntaxNode`s rather than `ISymbol`. I prefer to use syntax analysis whenever possible because I assume it is faster than semantic model-based analysis.